### PR TITLE
Add soft deletion for project submissions

### PR DESCRIPTION
**Pull request by Tenacious-Qi**
This addresses issue #1885

I hope I did this correctly. I implemented the 'discard' gem to handle soft deletion of flagged submissions. An admin can now remove a flagged submission and then see the flag in the resolved flags list.The "taken action" will be "removed_project_submission". I added a `RemoveSubmission` class to `app/services/admin/flags`, mimicking the structure of `BanUser` and `Dismiss`. 

To address the logic of "a user can only have one submission per project", I changed the index on `lesson_id` / `user_id`  of `project_submissions` to be a partial index, where the unique constraint only applies to submissions that have not been manually discarded by an admin. This means that a user can have their submission discarded by an admin and then later submit a project submission that is more appropriate. However, they still cannot submit more than one "non-discarded" submission to a project. 
I tested this in console and it still results in a `ActiveRecord::RecordNotUnique` error. 

Additionally, If an admin manually discards a submission, the user will no longer see their submission on the lesson page, and no user will be able to see a discarded submission ( added `discarded_at: nil` to the ProjectSubmission `viewable` scope). I have also added a few specs to test soft deletion.

### CODE REVIEW

**Comment by Tenacious-Qi** [comment]
This references issue #1885

File: app/admin/flags.rb [hunk 1]
```diff
@@ -87,9 +87,14 @@ def dismiss
     end
 
     def remove_project_submission
-      resource.project_submission.destroy
+      result = Admin::Flags::RemoveSubmission.call(current_user, flag: resource)
 
-      redirect_to admin_flags_path, notice: 'Success: Submission has been removed.'
+      if result.success?
+        resource.project_submission.discard
```
Comment (KevinMulhern) [inline app/admin/flags.rb:93]: Could we move this into the `RemoveSubmission` service please?

File: app/models/project_submission.rb [hunk 2]
```diff
@@ -1,4 +1,6 @@
 class ProjectSubmission < ApplicationRecord
```
Comment (KevinMulhern) [inline app/models/project_submission.rb:1]: I think we should add a default scope to the ProjectSubmissions model so all the discarded submissions are filtered out by default, that will allow us to remove the `discarded_at: nil` conditions on the viewable scope and in the lessons_helper.

Discard have an example of using a default scope here: https://github.com/jhawthorn/discard#default-scope

File: db/migrate/20210512034228_add_partial_index_to_lesson_id_user_id_project_submissions.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddPartialIndexToLessonIdUserIdProjectSubmissions < ActiveRecord::Migration[6.1]
+  def change
+    add_index :project_submissions, %i[user_id lesson_id], unique: true, where: 'discarded_at is null'
```
Comment (KevinMulhern) [inline db/migrate/20210512034228_add_partial_index_to_lesson_id_user_id_project_submissions.rb:3]: Using a partial index is a super slick way of solving the one submission per project problem 👏

File: spec/services/admin/flags/remove_submission_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+require 'rails_helper'
+
+RSpec.describe Admin::Flags::RemoveSubmission do
+  let(:admin) { create(:user, admin: true) }
```
Comment (KevinMulhern) [inline spec/services/admin/flags/remove_submission_spec.rb:4]: Just a nitpick, can we have the subject be up top with the two lets below please? 

```suggestion
  subject(:service) { described_class.call(admin, flag: flag) }
  
  let(:admin) { create(:user, admin: true) }
  let(:flag) { create(:flag) }
```

File: spec/system/admin/flags_spec.rb [hunk 5]
```diff
@@ -54,6 +54,12 @@
         expect(page).not_to have_content(submission_owner.username)
       end
     end
+
+    it 'shows the flag in the resolved list' do
+      visit '/admin/flags?scope=resolved'
```
Comment (KevinMulhern) [inline spec/system/admin/flags_spec.rb:59]: Nit pick, can we use a route helper here instead please?  `visit admin_flags_path(scope: 'resolved')` that will throw up a nice error if these urls ever change.

**Review by KevinMulhern — COMMENTED** [review]
Brilliant work @Tenacious-Qi 💪 

 Just a few little nitpicks, but overall this is super clean 💯

File: app/admin/flags.rb (hunk 1 above)
Comment (Tenacious-Qi) [reply app/admin/flags.rb]: Yep! That makes more sense there.

File: app/models/project_submission.rb (hunk 2 above)
Comment (Tenacious-Qi) [reply app/models/project_submission.rb:1]: Much better this way, thanks! I just kept it simple with `default_scope -> { kept }` in the submission model. I think that's good enough? Seems to be working the same. Now, if we want to find a discarded post, I guess we can do `ProjectSubmission.with_discarded`.

File: db/migrate/20210512034228_add_partial_index_to_lesson_id_user_id_project_submissions.rb (hunk 3 above)
Comment (Tenacious-Qi) [reply db/migrate/20210512034228_add_partial_index_to_lesson_id_user_id_project_submissions.rb:3]: :smile:  I was happy when I found this! I googled a bunch and didn't find much, and then there it was in the Rails docs haha.

File: spec/services/admin/flags/remove_submission_spec.rb (hunk 4 above)
Comment (Tenacious-Qi) [reply spec/services/admin/flags/remove_submission_spec.rb]: Yep definitely looks better that way. I noticed that `Dismiss` and `BanUser` also did not have `subject` at the top, so most recent commit also changes those. Hope that's okay.

File: spec/system/admin/flags_spec.rb (hunk 5 above)
Comment (Tenacious-Qi) [reply spec/system/admin/flags_spec.rb]: done :+1:

File: app/models/project_submission.rb (hunk 2 above)
Comment (KevinMulhern) [reply app/models/project_submission.rb:1]: Perfect 👌

**Review by KevinMulhern — APPROVED** [review]
Looks great to me @Tenacious-Qi Thank you for all your work on this, its great to have this properly working now 🥳

**Pull request merged**